SPI Driver

Novakov opened the pull request:
* SPI Driver using DMA for transfers
* RAII-style slave selection
* Reworked io_map in much more readable (and strongly typed) manner 
* Compile-type verification that pins are not overlapping 
  * EBI configuration skipped for now as it is a bit tricky

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pw-sat2/pwsat2obc/64)
<!-- Reviewable:end -->

Novakov commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[SPI] Implement new driver](https://trello.com/c/t2AKU0KA/95-spi-implement-new-driver)

zaq32 commented:
Reviewed 45 of 69 files at r1.
Review status: 44 of 66 files reviewed at latest revision, 12 unresolved discussions.

---

*[integration_tests/obc/obc.py, line 40 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_Vi9fjimkkb2IOWvDD:-K_Vi9fjimkkb2IOWvDE:bqnz7ns) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/integration_tests/obc/obc.py#L40)):*
> ```Python
>         response = self._terminal.command("getState")
>         while response != "1":
>             # time.sleep(5)
> ```

if it is not needed remove it

---

*[libs/base/Include/utils.h, line 76 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_ViO1TeLt1XpX2iMuL:-K_ViO1TeLt1XpX2iMuM:b3gav6k) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/base/Include/utils.h#L76)):*
> ```objc
>  * @return Value in Hz
>  */
> constexpr std::uint32_t operator"" _MHz(unsigned long long int value)
> ```

in order to make it something more than syntactic sugar dedicated type would be preferable than raw integer

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 40 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VjuiEMbTaDhiEGxHJ:-K_VjuiEMbTaDhiEGxHK:be4v0h2) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L40)):*
> ```objc
>             inline void Low() const;
>             /** @brief Toggles pin */
>             inline void Toggle() const;
> ```

Is it allowed to call these functions on input pin?

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 45 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VjmJo0_BQ1OjH1XvZ:-K_VjmJo0_BQ1OjH1Xv_:blwuh3) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L45)):*
> ```objc
>              * @return true if pin is high
>              */
>             inline bool Input() const;
> ```

is it allowed to call this function on output pin?

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 51 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VjGGWpwREKPLuqMcU:-K_VjGGWpwREKPLuqMcV:b-xed0sh) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L51)):*
> ```objc
>             const GPIO_Port_TypeDef _port;
>             /** @brief Pin */
>             const std::uint16_t _pin;
> ```

these members should not be protected. If they should be accessible only in derived classes they should be accessible via dedicated protected interface. 

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 79 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VkA3M3QUUlfsQe8OF:-K_VkA3M3QUUlfsQe8OG:b-39ujmq) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L79)):*
> ```objc
>          * @tparam DefaultState true if pin should be high by default
>          */
>         template <typename Location, bool DefaultState = true> class OutputPin final : public Pin
> ```

I'd recommend hiding Pin as private class member where its features can be controlled instead of public base class.

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 100 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_Vj_enjTCP14IowL62:-K_Vj_enjTCP14IowL63:bcqz319) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L100)):*
> ```objc
>         template <typename Location, bool DefaultState> void OutputPin<Location, DefaultState>::Initialize() const
>         {
>             GPIO_PinModeSet(this->_port, this->_pin, gpioModePushPull, ToInt(DefaultState));
> ```

why use members instead of template parameters?

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 108 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VkJQdQ5KFFTk0_YE3:-K_VkJQdQ5KFFTk0_YE4:b-39ujmq) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L108)):*
> ```objc
>         * @tparam DefaultState true if pin pull up should be enabled
>         */
>         template <typename Location, bool DefaultState = true> class InputPin final : public Pin
> ```

I'd recommend hiding Pin as private class member where its features can be controlled instead of public base class.

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 129 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VjgfXyGS9XJtYMuTk:-K_VjgfXyGS9XJtYMuTl:bcqz319) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L129)):*
> ```objc
>         template <typename Location, bool DefaultState> void InputPin<Location, DefaultState>::Initialize() const
>         {
>             GPIO_PinModeSet(this->_port, this->_pin, gpioModeInputPull, ToInt(DefaultState));
> ```

why use members instead of template parameters?

---

*[libs/drivers/leuart/leuart.cpp, line 50 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VkUv9bMG0mr6ZOnxF:-K_VkUv9bMG0mr6ZOnxG:b-cjrwna) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/leuart/leuart.cpp#L50)):*
> ```cpp
> 
>     //    GPIO_PinModeSet(LEUART0_PORT, LEUART0_TX, gpioModePushPull, 1);
>     //    GPIO_PinModeSet(LEUART0_PORT, LEUART0_RX, gpioModeInputPull, 1);
> ```

if those line are not  needed anymore remove them

---

*[libs/drivers/n25q/n25q.cpp, line 317 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VvWbEymfMXQcvCmeN:-K_VvWbEymfMXQcvCmeO:b-ojnrcb) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/n25q/n25q.cpp#L317)):*
> ```cpp
>         if (timeoutCheck.Expired())
>             return OperationResult::Timeout;
>     } while (true);
> ```

sleep/yield would at the end of a loop would be nice

---

*[libs/efm_support/CMakeLists.txt, line 1 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VwsEvuXpiKVurDQxH:-K_VwsEvuXpiKVurDQxI:bz96qa5) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/efm_support/CMakeLists.txt#L1)):*
> ```CMake
> set(NAME efm_support)
> ```

what is the purpose of this library?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64)*
<!-- Sent from Reviewable.io -->

zaq32 commented:
Reviewed 20 of 69 files at r1, 3 of 6 files at r2, 9 of 9 files at r3.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[libs/base/Include/base/io_map.h, line 61 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_ysnyRwogKAdKkbZbt:-K_ysnyRwogKAdKkbZbu:b-7khyww) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/base/Include/base/io_map.h#L61)):*
> ```objc
>      * @tparam TPinNumber Pin number
>      */
>     template <GPIO_Port_TypeDef TPort, std::uint16_t TPinNumber> struct PinLocation : public PinTag
> ```

missing include: cstdint

---

*[libs/drivers/spi/Include/spi/efm.h, line 27 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_z_eUMOeHPcROnP3Jg:-K_z_eUMOeHPcROnP3Jh:bvmtmw0) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/drivers/spi/Include/spi/efm.h#L27)):*
> ```objc
>          * This class represent EFM peripheral, to use it with specific device (slave) see @ref EFMSPISlaveInterface
>          */
>         class EFMSPIInterface final
> ```

There is some non obvious logic inside this class that is not tested at all. 

---

*[libs/efm_support/Include/efm_support/clock.h, line 8 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zaQH9tTUzq6D5P1L1:-K_zaQH9tTUzq6D5P1L2:b-kje888) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/efm_support/Include/efm_support/clock.h#L8)):*
> ```objc
> namespace efm
> {
>     template <typename Peripheral> constexpr CMU_Clock_TypeDef Clock(const Peripheral hw)
> ```

undocumented

---

*[libs/efm_support/Include/efm_support/dma.h, line 16 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zaT_PDS8zp7iq72ry:-K_zaT_PDS8zp7iq72rz:b-kje888) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/efm_support/Include/efm_support/dma.h#L16)):*
> ```objc
>     };
> 
>     template <DMASignalUSART Signal> constexpr DMADRV_PeripheralSignal_t DMASignal(const USART_TypeDef* peripheral) = delete;
> ```

undocumented

---

*[libs/obc/hardware/Include/obc/gpio-internal.h, line 21 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zSbrjBLbDC7hB7cVq:-K_zSbrjBLbDC7hB7cVr:b-326k21) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/hardware/Include/obc/gpio-internal.h#L21)):*
> ```objc
> 
>         /** @brief Conflict detection */
>         struct Conflicts
> ```

The overall mechanism works as expected but this mechanism leaves something to be desired. As any new ping set (i.e. cameras, additional chip select pins) will require updating this implementation. Ideally it should not be aware of any specific pin group type and operate either on generic pin group type like this:

```
template <typename... PinLocations> struct PinGroup
{
/* (...) */
};
```
or pure flat set of pin locations.

Plus it is no immediately obvious that this type works only on limited set of tagged types.

---

*[libs/obc/hardware/Include/obc/gpio.h, line 83 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_z6jzNu0_jFI_cjh4c:-K_z6jzNu0_jFI_cjh4d:b-iv7iqt) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/hardware/Include/obc/gpio.h#L83)):*
> ```objc
>     {
>         /** @brief Slave Select - Flash1 */
>         const drivers::gpio::OutputPin<TSlaveSelectFlash1> SlaveSelectFlash1;
> ```

I'd call it flash1ChipSelect

---

*[libs/obc/storage/n25q_flash/n25q.cpp, line 19 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zTsYX0O3a2N760VD3:-K_zTsYX0O3a2N760VD4:b-4g6fwy) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/storage/n25q_flash/n25q.cpp#L19)):*
> ```cpp
> void N25QStorage::Initialize()
> {
>     this->ExternalFlashDriver.Reset();
> ```

Does this function guarantees that  by the time it returns the chip will already be in operational state?

---

*[libs/obc/storage/n25q_flash/Include/obc/storage/n25q.h, line 24 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zY8Bbm8OiN9IMRa6T:-K_zY8Bbm8OiN9IMRa6U:b-5s965j) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/storage/n25q_flash/Include/obc/storage/n25q.h#L24)):*
> ```objc
>          * @brief Storage driver for N25Q
>          */
>         class N25QStorage final
> ```

it would be nice to have documented the purpose of this class. as it takes a while to understand it and to be strict this class is not a driver...

---

*[libs/obc/storage/n25q_flash/Include/obc/storage/n25q.h, line 54 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zXDadHQOf4V5svRuR:-K_zXDaerpzxgAz7o-H9:b-fduyz1) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/storage/n25q_flash/Include/obc/storage/n25q.h#L54)):*
> ```objc
> 
>             /** @brief N25Q Yaffs device */
>             devices::n25q::N25QYaffsDevice<devices::n25q::BlockMapping::Sector, 512_Bytes, 16_MB> ExternalFlash;
> ```

why these fields are public? They don't seem to be used outside this class.

---

*[libs/terminal/terminal.cpp, line 101 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zWZ4xT_MosBsjbeYt:-K_zWZ4xT_MosBsjbeYu:bt2dyb2) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/terminal/terminal.cpp#L101)):*
> ```cpp
>         terminal->_stdio.Readline(&terminal->_stdio, input_buffer, COUNT_OF(input_buffer));
> 
>         //        LOGF(LOG_LEVEL_INFO, "Received line %s", input_buffer);
> ```

either remove this line or uncomment it.

---

*[src/main.cpp, line 55 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zUyR2MWf44gFytfAx:-K_zUyR2MWf44gFytfAy:bdz1xm3) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/src/main.cpp#L55)):*
> ```cpp
> extern "C" void vApplicationIdleHook(void)
> {
>     //    EMU_EnterEM1();
> ```

why disable EM1?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64)*
<!-- Sent from Reviewable.io -->

Novakov commented:
Review status: 52 of 66 files reviewed at latest revision, 23 unresolved discussions.

---

*[integration_tests/obc/obc.py, line 40 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_Vi9fjimkkb2IOWvDD:-K_njnWMs3fbphC-hljV:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/integration_tests/obc/obc.py#L40)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

if it is not needed remove it
</blockquote></details>

Done.

---

*[libs/base/Include/utils.h, line 76 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_ViO1TeLt1XpX2iMuL:-K_WAnbP3RXzE7918mBo:b-lp3vsk) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/base/Include/utils.h#L76)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

in order to make it something more than syntactic sugar dedicated type would be preferable than raw integer
</blockquote></details>

It's just syntactic sugar too avoid counting trailing zeros in numbers

---

*[libs/base/Include/base/io_map.h, line 61 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_ysnyRwogKAdKkbZbt:-K_zlre4fFe9h2dtAh_p:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/base/Include/base/io_map.h#L61)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

missing include: cstdint
</blockquote></details>

Done.

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 40 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VjuiEMbTaDhiEGxHJ:-K_WBVEMxJOgBUdxdhE9:b40p40o) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L40)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

Is it allowed to call these functions on input pin?
</blockquote></details>

Sure - it will enable/disable/toggle pull-up

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 45 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VjmJo0_BQ1OjH1XvZ:-K_WBkeaZXLDo4ZDEZvW:b-lxnlt5) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L45)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

is it allowed to call this function on output pin?
</blockquote></details>

Sure - it will give pin value 

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 51 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VjGGWpwREKPLuqMcU:-K_nkuvuk9bR4NCKbMWL:b-fzvzgx) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L51)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

these members should not be protected. If they should be accessible only in derived classes they should be accessible via dedicated protected interface. 
</blockquote></details>

Changed to private

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 79 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VkA3M3QUUlfsQe8OF:-K_nlVt5MEmQNIcwr5wE:b-wze6yr) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L79)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

I'd recommend hiding Pin as private class member where its features can be controlled instead of public base class.
</blockquote></details>

We discussed that and decided to leave as it is now

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 100 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_Vj_enjTCP14IowL62:-K_nlkEjTrU2_bN1zqOA:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L100)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

why use members instead of template parameters?
</blockquote></details>

Done.

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 108 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VkJQdQ5KFFTk0_YE3:-K_nlqFf0f38hBNpMCQN:b-wze6yr) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L108)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

I'd recommend hiding Pin as private class member where its features can be controlled instead of public base class.
</blockquote></details>

We discussed that and decided to leave as it is now

---

*[libs/drivers/gpio/Include/gpio/gpio.h, line 129 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VjgfXyGS9XJtYMuTk:-K_nlls2Lpnrmqnuu8PB:b-896fjx) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/gpio/Include/gpio/gpio.h#L129)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

why use members instead of template parameters?
</blockquote></details>

Done

---

*[libs/drivers/leuart/leuart.cpp, line 50 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VkUv9bMG0mr6ZOnxF:-K_nluPlDPzdmHFA2Vue:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/leuart/leuart.cpp#L50)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

if those line are not  needed anymore remove them
</blockquote></details>

Done.

---

*[libs/drivers/n25q/n25q.cpp, line 317 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VvWbEymfMXQcvCmeN:-K_nm38PkIa0HVBEUGyQ:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/drivers/n25q/n25q.cpp#L317)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

sleep/yield would at the end of a loop would be nice
</blockquote></details>

Done.

---

*[libs/drivers/spi/Include/spi/efm.h, line 27 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_z_eUMOeHPcROnP3Jg:-K_zx07fX1mf_xfZSnXe:b79gyer) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/drivers/spi/Include/spi/efm.h#L27)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

There is some non obvious logic inside this class that is not tested at all. 
</blockquote></details>

How do you propose to unit test this?

To decouple SPI driver from hardware will require creating wrapper around emlib and dmadrv .

---

*[libs/efm_support/CMakeLists.txt, line 1 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_VwsEvuXpiKVurDQxH:-K_WC6MUXcueZzMMRYpH:btpnndb) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/995bd3c0f2631e4358cd482260258614a02cef3d/libs/efm_support/CMakeLists.txt#L1)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

what is the purpose of this library?
</blockquote></details>

When working with some periperhal (like USART1) you need to know few another constant/registers like clock (cmuClock_USART1), DMA signals (USART1_TXBL). peripheral is configured in io_map but putting there all these constants will just make that file less readable, so I created few constexprs to resolve that dependant constants. I can't put is in emlib as another file as some constants (DMA signals) comes from emdrv

---

*[libs/efm_support/Include/efm_support/clock.h, line 8 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zaQH9tTUzq6D5P1L1:-K_zluaflziK9B3zlJAc:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/efm_support/Include/efm_support/clock.h#L8)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

undocumented
</blockquote></details>

Done.

---

*[libs/efm_support/Include/efm_support/dma.h, line 16 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zaT_PDS8zp7iq72ry:-K_zkl6Hn2gvCNrQ36TX:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/efm_support/Include/efm_support/dma.h#L16)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

undocumented
</blockquote></details>

Done.

---

*[libs/obc/hardware/Include/obc/gpio-internal.h, line 21 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zSbrjBLbDC7hB7cVq:-K_zt6dN1RZE3j9xs-h2:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/hardware/Include/obc/gpio-internal.h#L21)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

The overall mechanism works as expected but this mechanism leaves something to be desired. As any new ping set (i.e. cameras, additional chip select pins) will require updating this implementation. Ideally it should not be aware of any specific pin group type and operate either on generic pin group type like this:

```
template <typename... PinLocations> struct PinGroup
{
/* (...) */
};
```
or pure flat set of pin locations.

Plus it is no immediately obvious that this type works only on limited set of tagged types.
</blockquote></details>

Done.

---

*[libs/obc/hardware/Include/obc/gpio.h, line 83 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_z6jzNu0_jFI_cjh4c:-K_zlAgc5B_c5xRFWU0j:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/hardware/Include/obc/gpio.h#L83)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

I'd call it flash1ChipSelect
</blockquote></details>

Done.

---

*[libs/obc/storage/n25q_flash/n25q.cpp, line 19 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zTsYX0O3a2N760VD3:-K_zwq49bARGaGfOHaZY:bxpo5xm) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/storage/n25q_flash/n25q.cpp#L19)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

Does this function guarantees that  by the time it returns the chip will already be in operational state?
</blockquote></details>

Added operation result as return value

---

*[libs/obc/storage/n25q_flash/Include/obc/storage/n25q.h, line 24 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zY8Bbm8OiN9IMRa6T:-K_zlPVNdMaeZbD89s7S:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/storage/n25q_flash/Include/obc/storage/n25q.h#L24)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

it would be nice to have documented the purpose of this class. as it takes a while to understand it and to be strict this class is not a driver...
</blockquote></details>

Done.

---

*[libs/obc/storage/n25q_flash/Include/obc/storage/n25q.h, line 54 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zXDadHQOf4V5svRuR:-K_zlZ70rFdJvNpcavGw:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/storage/n25q_flash/Include/obc/storage/n25q.h#L54)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

why these fields are public? They don't seem to be used outside this class.
</blockquote></details>

Done.

---

*[libs/terminal/terminal.cpp, line 101 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zWZ4xT_MosBsjbeYt:-K_zlc1f7s5Hlq8sgEPE:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/terminal/terminal.cpp#L101)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

either remove this line or uncomment it.
</blockquote></details>

Done.

---

*[src/main.cpp, line 55 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zUyR2MWf44gFytfAx:-K_zleB-18HeaXPKcwHw:b-pa0r7y) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/src/main.cpp#L55)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

why disable EM1?
</blockquote></details>

Debugging leftover

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64)*
<!-- Sent from Reviewable.io -->

zaq32 commented:
Reviewed 11 of 14 files at r4.
Review status: 63 of 66 files reviewed at latest revision, 6 unresolved discussions.

---

*[libs/drivers/spi/Include/spi/efm.h, line 27 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_z_eUMOeHPcROnP3Jg:-Ka-GZXcwk5hl7capbnN:bdni78n) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/drivers/spi/Include/spi/efm.h#L27)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

How do you propose to unit test this?

To decouple SPI driver from hardware will require creating wrapper around emlib and dmadrv .
</blockquote></details>

exactly

---

*[libs/efm_support/Include/efm_support/clock.h, line 13 at r4](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-Ka-GqqxT4UZ550SwkTA:-Ka-GqqxT4UZ550SwkTB:b587u36) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/db692f9c681e53ca4d683e3bae3b1cdbee14148f/libs/efm_support/Include/efm_support/clock.h#L13)):*
> ```objc
>      * @return CMU clock
>      *
>      * @remark Not all peripheral are supported at the moment - feel free to add new ones if needed
> ```

peripheral**s**

---

*[libs/obc/storage/n25q_flash/n25q.cpp, line 19 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zTsYX0O3a2N760VD3:-Ka-I6YbE2rcpuIhTrx0:b-39u3if) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/storage/n25q_flash/n25q.cpp#L19)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

Added operation result as return value
</blockquote></details>

Operation result  is one step but my point here is that device reset can take i.e. 10ms. Will the ```Reset``` procedure wait that 10ms until the device finishes its restart procedure or will it simply trigger the process & return?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64)*
<!-- Sent from Reviewable.io -->

Novakov commented:
Review status: 49 of 76 files reviewed at latest revision, 6 unresolved discussions.

---

*[libs/drivers/spi/Include/spi/efm.h, line 27 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_z_eUMOeHPcROnP3Jg:-Ka8PF1wOz9yGG3nAErl:b-ep54n1) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/drivers/spi/Include/spi/efm.h#L27)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

exactly
</blockquote></details>

I've added some basic tests. 

---

*[libs/efm_support/Include/efm_support/clock.h, line 13 at r4](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-Ka-GqqxT4UZ550SwkTA:-Ka3AX6IIs_jRCUa-XzT:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/db692f9c681e53ca4d683e3bae3b1cdbee14148f/libs/efm_support/Include/efm_support/clock.h#L13)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

peripheral**s**
</blockquote></details>

Done.

---

*[libs/obc/storage/n25q_flash/n25q.cpp, line 19 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zTsYX0O3a2N760VD3:-Ka14FuT4I02PhT7xb-z:bfmdmv2) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/storage/n25q_flash/n25q.cpp#L19)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

Operation result  is one step but my point here is that device reset can take i.e. 10ms. Will the ```Reset``` procedure wait that 10ms until the device finishes its restart procedure or will it simply trigger the process & return?
</blockquote></details>

It will wait until device is operational (or fail with timeout)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64)*
<!-- Sent from Reviewable.io -->

zaq32 commented:
Reviewed 3 of 14 files at r4, 24 of 24 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[libs/obc/hardware/Include/obc/gpio-internal.h, line 21 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zSbrjBLbDC7hB7cVq:-Ka8rBTrkq1n2RUfRPda:b-t9ewpg) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/hardware/Include/obc/gpio-internal.h#L21)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This much better but it can be further simplified by splatting the Pin, PinGroup, PinContainer collection into flat list of the Pins and then iterating over them in the way it is already done. This way you can remove Type/GetType elements and significantly simplify Conflict type.

---

*[libs/obc/storage/n25q_flash/n25q.cpp, line 19 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zTsYX0O3a2N760VD3:-Ka8ovdMublfH7UnoGmT:b-mjjk8v) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/storage/n25q_flash/n25q.cpp#L19)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

It will wait until device is operational (or fail with timeout)
</blockquote></details>

this fact should be documented

---

*[unit_tests/mock/emlib.h, line 139 at r5](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-Ka8qRlzQ9Lh5cfgan2v:-Ka8qRlzQ9Lh5cfgan2w:b-dkqfvs) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/dcd388aa751f86a89fbb57edf84d1d8947344f64/unit_tests/mock/emlib.h#L139)):*
>     this->_released = tmp._released;
>     this->_holder = tmp._holder;
This is conceptual resource leak. The right way here is to simply:
```
using std::swap;
swap(this->_released, tmp._released);
swap(this->_holder, tmp._holder);
```

or better (if the class supports swapping)
```
using std::swap;
swap(*this, tmp);
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64)*
<!-- Sent from Reviewable.io -->

Novakov commented:
Review status: 73 of 76 files reviewed at latest revision, 3 unresolved discussions.

---

*[libs/obc/hardware/Include/obc/gpio-internal.h, line 21 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zSbrjBLbDC7hB7cVq:-KaDpc0pumOkhKKl5BYl:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/hardware/Include/obc/gpio-internal.h#L21)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

This much better but it can be further simplified by splatting the Pin, PinGroup, PinContainer collection into flat list of the Pins and then iterating over them in the way it is already done. This way you can remove Type/GetType elements and significantly simplify Conflict type.
</blockquote></details>

Done.

---

*[libs/obc/storage/n25q_flash/n25q.cpp, line 19 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-K_zTsYX0O3a2N760VD3:-Ka9BFYSBDzNmLrB7WWZ:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/be1f8e5ef08d5c372e877029c8a549e5ee4c2fc1/libs/obc/storage/n25q_flash/n25q.cpp#L19)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

this fact should be documented
</blockquote></details>

Done.

---

*[unit_tests/mock/emlib.h, line 139 at r5](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64#-Ka8qRlzQ9Lh5cfgan2v:-Ka9BEKVHTn4TvrHC52G:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/dcd388aa751f86a89fbb57edf84d1d8947344f64/unit_tests/mock/emlib.h#L139)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

>     this->_released = tmp._released;
>     this->_holder = tmp._holder;
This is conceptual resource leak. The right way here is to simply:
```
using std::swap;
swap(this->_released, tmp._released);
swap(this->_holder, tmp._holder);
```

or better (if the class supports swapping)
```
using std::swap;
swap(*this, tmp);
```


</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64)*
<!-- Sent from Reviewable.io -->

zaq32 commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/64)*
<!-- Sent from Reviewable.io -->
